File: neonfs-cli/src/commands/dr.rs
//! DR snapshot commands
//!
//! Operator-facing surface for the cluster's disaster-recovery snapshot
//! pipeline. Each snapshot captures `MetadataStateMachine` indexes plus
//! cluster CA material at one Ra log point — see
//! `NeonFS.Core.DRSnapshot` for the on-disk format.
//!
//! The handlers on the daemon side live in
//! `NeonFS.CLI.Handler.handle_dr_snapshot_*`.

use crate::daemon::DaemonConnection;
use crate::error::{CliError, Result};
use crate::output::{json, table, OutputFormat};
use crate::term::{
    extract_error, term_to_list, term_to_map, term_to_string, term_to_u64, unwrap_ok_tuple,
};
use clap::Subcommand;
use eetf::{Binary, Map, Term};
use serde::Serialize;
use std::collections::BTreeMap;

/// DR snapshot subcommands
#[derive(Debug, Subcommand)]
pub enum DrCommand {
    /// Snapshot management
    Snapshot {
        #[command(subcommand)]
        command: DrSnapshotCommand,
    },

    /// Full-cluster restore (#1005): stage + apply an exported DR
    /// snapshot, then restore each volume's content from its backup
    /// archive. Run this on a freshly-bootstrapped single node
    /// (`neonfs cluster init`); reattach the remaining nodes with
    /// `neonfs cluster join` afterwards.
    Restore {
        /// Source directory produced by `dr snapshot export`, holding the
        /// snapshot plus (by default) `volumes/<name>.backup` archives.
        #[arg(long)]
        source: String,

        /// Optional JSON catalogue `{"<volume>": "<archive-path>"}`
        /// pinning where each volume's backup archive lives. Relative
        /// paths resolve against `--source`; the catalogue is
        /// authoritative, so omitted volumes are left as empty shells.
        #[arg(long)]
        catalogue: Option<String>,

        /// Passphrase for encrypted backup archives (#1004).
        #[arg(long)]
        passphrase: Option<String>,
    },
}

/// Snapshot subcommand variants
#[derive(Debug, Subcommand)]
pub enum DrSnapshotCommand {
    /// Create an immediate snapshot of the cluster's metadata + CA state
    Create {
        /// Optional human-readable label (recorded in the audit log only)
        #[arg(long)]
        label: Option<String>,
    },

    /// List every snapshot in the `_system` volume's `/dr` directory
    List,

    /// Show a single snapshot's manifest
    Show {
        /// Snapshot ID (the timestamp directory under `/dr`)
        id: String,
    },

    /// Apply a snapshot's cluster-wide metadata back into live Ra
    /// state (#1005). Overlays the eight cluster-wide keyspaces
    /// (volumes, services, encryption keys, ACLs, segment
    /// assignments, credentials, escalations, KV); per-volume content
    /// is restored separately via `backup restore`.
    Apply {
        /// Snapshot ID (the timestamp directory under `/dr`)
        id: String,
    },

    /// Export a snapshot off-cluster to a directory on the daemon's
    /// filesystem (#1367) so it survives a bare-metal disaster — the
    /// in-cluster copy is destroyed with `_system`.
    Export {
        /// Snapshot ID (the timestamp directory under `/dr`)
        id: String,

        /// Destination directory on the daemon's filesystem
        #[arg(long)]
        out: String,
    },

    /// Stage an exported snapshot back into a freshly-bootstrapped
    /// cluster's `_system` volume (#1367), ready for `apply`.
    Import {
        /// Source directory produced by `dr snapshot export`
        source: String,
    },
}

#[derive(Debug, Serialize)]
struct DrRestoreResult {
    id: String,
    total: u64,
    generation: u64,
    restored: BTreeMap<String, u64>,
}

impl DrRestoreResult {
    fn from_term(term: Term) -> Result<Self> {
        let map = term_to_map(&term)?;

        let id = term_to_string(
            map.get("id")
                .ok_or_else(|| CliError::TermConversionError("Missing 'id' field".to_string()))?,
        )?;

        let total = map
            .get("total")
            .and_then(|t| term_to_u64(t).ok())
            .unwrap_or(0);
        let generation = map
            .get("generation")
            .and_then(|t| term_to_u64(t).ok())
            .unwrap_or(0);

        let restored = map
            .get("restored")
            .and_then(|t| term_to_map(t).ok())
            .map(|m| {
                m.iter()
                    .filter_map(|(k, v)| term_to_u64(v).ok().map(|n| (k.clone(), n)))
                    .collect()
            })
            .unwrap_or_default();

        Ok(Self {
            id,
            total,
            generation,
            restored,
        })
    }
}

#[derive(Debug, Serialize)]
struct DrVolumeRestore {
    name: String,
    status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    archive: Option<String>,
    files: u64,
    bytes: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    reason: Option<String>,
}

impl DrVolumeRestore {
    fn from_term(term: &Term) -> Result<Self> {
        let map = term_to_map(term)?;
        Ok(Self {
            name: map
                .get("name")
                .map(term_to_string)
                .transpose()?
                .unwrap_or_default(),
            status: map
                .get("status")
                .map(term_to_string)
                .transpose()?
                .unwrap_or_default(),
            archive: map.get("archive").map(term_to_string).transpose()?,
            files: map
                .get("files")
                .and_then(|t| term_to_u64(t).ok())
                .unwrap_or(0),
            bytes: map
                .get("bytes")
                .and_then(|t| term_to_u64(t).ok())
                .unwrap_or(0),
            reason: map.get("reason").map(term_to_string).transpose()?,
        })
    }
}

#[derive(Debug, Serialize)]
struct DrFullRestoreResult {
    snapshot_id: String,
    generation: u64,
    restored: BTreeMap<String, u64>,
    volumes: Vec<DrVolumeRestore>,
    volumes_restored: u64,
    volumes_failed: u64,
}

impl DrFullRestoreResult {
    fn from_term(term: Term) -> Result<Self> {
        let map = term_to_map(&term)?;

        let snapshot_id = term_to_string(map.get("snapshot_id").ok_or_else(|| {
            CliError::TermConversionError("Missing 'snapshot_id' field".to_string())
        })?)?;

        let generation = map
            .get("generation")
            .and_then(|t| term_to_u64(t).ok())
            .unwrap_or(0);

        let restored = map
            .get("restored")
            .and_then(|t| term_to_map(t).ok())
            .map(|m| {
                m.iter()
                    .filter_map(|(k, v)| term_to_u64(v).ok().map(|n| (k.clone(), n)))
                    .collect()
            })
            .unwrap_or_default();

        let volumes = map
            .get("volumes")
            .map(term_to_list)
            .transpose()?
            .unwrap_or_default()
            .iter()
            .map(DrVolumeRestore::from_term)
            .collect::<Result<Vec<_>>>()?;

        let volumes_restored = map
            .get("volumes_restored")
            .and_then(|t| term_to_u64(t).ok())
            .unwrap_or(0);
        let volumes_failed = map
            .get("volumes_failed")
            .and_then(|t| term_to_u64(t).ok())
            .unwrap_or(0);

        Ok(Self {
            snapshot_id,
            generation,
            restored,
            volumes,
            volumes_restored,
            volumes_failed,
        })
    }
}

#[derive(Debug, Serialize)]
struct DrSnapshotEntry {
    id: String,
    path: String,
    created_at: String,
    state_version: Option<u64>,
    file_count: u64,
    total_bytes: u64,
}

impl DrSnapshotEntry {
    fn from_term(term: Term) -> Result<Self> {
        let map = term_to_map(&term)?;

        let id = term_to_string(
            map.get("id")
                .ok_or_else(|| CliError::TermConversionError("Missing 'id' field".to_string()))?,
        )?;

        let path =
            term_to_string(map.get("path").ok_or_else(|| {
                CliError::TermConversionError("Missing 'path' field".to_string())
            })?)?;

        let created_at = term_to_string(map.get("created_at").ok_or_else(|| {
            CliError::TermConversionError("Missing 'created_at' field".to_string())
        })?)?;

        let state_version = map.get("state_version").and_then(|t| term_to_u64(t).ok());
        let file_count = map
            .get("file_count")
            .and_then(|t| term_to_u64(t).ok())
            .unwrap_or(0);
        let total_bytes = map
            .get("total_bytes")
            .and_then(|t| term_to_u64(t).ok())
            .unwrap_or(0);

        Ok(Self {
            id,
            path,
            created_at,
            state_version,
            file_count,
            total_bytes,
        })
    }

    fn id_short(&self) -> String {
        if self.id.len() > 17 {
            format!("{}…", &self.id[..16])
        } else {
            self.id.clone()
        }
    }
}

impl DrCommand {
    pub fn execute(&self, format: OutputFormat) -> Result<()> {
        match self {
            DrCommand::Snapshot { command } => command.execute(format),
            DrCommand::Restore {
                source,
                catalogue,
                passphrase,
            } => Self::restore(source, catalogue.as_deref(), passphrase.as_deref(), format),
        }
    }

    fn restore(
        source: &str,
        catalogue: Option<&str>,
        passphrase: Option<&str>,
        format: OutputFormat,
    ) -> Result<()> {
        let mut opts = vec![];
        if let Some(catalogue) = catalogue {
            opts.push((binary("catalogue"), binary(catalogue)));
        }
        if let Some(passphrase) = passphrase {
            opts.push((binary("passphrase"), binary(passphrase)));
        }
        let opts_term = Term::Map(Map {
            map: opts.into_iter().collect(),
        });

        let result = smol::block_on(async {
            let mut conn = DaemonConnection::connect().await?;
            conn.call(
                "Elixir.NeonFS.CLI.Handler",
                "handle_dr_restore",
                vec![binary(source), opts_term],
            )
            .await
        })?;

        if let Some(err) = extract_error(&result) {
            return Err(err);
        }

        let restore = DrFullRestoreResult::from_term(unwrap_ok_tuple(result)?)?;

        match format {
            OutputFormat::Json => println!("{}", json::format(&restore)?),
            OutputFormat::Table => {
                println!("Restored cluster from snapshot {}", restore.snapshot_id);
                println!("  Cluster generation: {}", restore.generation);
                println!(
                    "  Volumes restored:   {} ({} failed)",
                    restore.volumes_restored, restore.volumes_failed
                );
                for volume in &restore.volumes {
                    match volume.status.as_str() {
                        "restored" => println!(
                            "    ✓ {:<24} {} file(s), {} byte(s)",
                            volume.name, volume.files, volume.bytes
                        ),
                        "skipped" => println!(
                            "    - {:<24} skipped ({})",
                            volume.name,
                            volume.reason.as_deref().unwrap_or("no archive")
                        ),
                        _ => println!(
                            "    ✗ {:<24} failed ({})",
                            volume.name,
                            volume.reason.as_deref().unwrap_or("unknown")
                        ),
                    }
                }
            }
        }

        Ok(())
    }
}

impl DrSnapshotCommand {
    pub fn execute(&self, format: OutputFormat) -> Result<()> {
        match self {
            DrSnapshotCommand::Create { label } => Self::create(label.as_deref(), format),
            DrSnapshotCommand::List => Self::list(format),
            DrSnapshotCommand::Show { id } => Self::show(id, format),
            DrSnapshotCommand::Apply { id } => Self::apply(id, format),
            DrSnapshotCommand::Export { id, out } => Self::export(id, out, format),
            DrSnapshotCommand::Import { source } => Self::import(source, format),
        }
    }

    fn export(id: &str, out: &str, format: OutputFormat) -> Result<()> {
        let result = smol::block_on(async {
            let mut conn = DaemonConnection::connect().await?;
            conn.call(
                "Elixir.NeonFS.CLI.Handler",
                "handle_dr_snapshot_export",
                vec![binary(id), binary(out)],
            )
            .await
        })?;

        if let Some(err) = extract_error(&result) {
            return Err(err);
        }

        let map = term_to_map(&unwrap_ok_tuple(result)?)?;
        let dest = map
            .get("dest")
            .map(term_to_string)
            .transpose()?
            .unwrap_or_default();
        let files = map
            .get("file_count")
            .map(term_to_u64)
            .transpose()?
            .unwrap_or(0);

        match format {
            OutputFormat::Json => {
                println!(
                    "{{\"id\":\"{}\",\"dest\":\"{}\",\"file_count\":{}}}",
                    id, dest, files
                )
            }
            OutputFormat::Table => {
                println!("Exported snapshot {}", id);
                println!("  Destination: {}", dest);
                println!("  Files:       {}", files);
            }
        }

        Ok(())
    }

    fn import(source: &str, format: OutputFormat) -> Result<()> {
        let result = smol::block_on(async {
            let mut conn = DaemonConnection::connect().await?;
            conn.call(
                "Elixir.NeonFS.CLI.Handler",
                "handle_dr_snapshot_import",
                vec![binary(source)],
            )
            .await
        })?;

        if let Some(err) = extract_error(&result) {
            return Err(err);
        }

        let map = term_to_map(&unwrap_ok_tuple(result)?)?;
        let id = map
            .get("id")
            .map(term_to_string)
            .transpose()?
            .unwrap_or_default();
        let files = map
            .get("file_count")
            .map(term_to_u64)
            .transpose()?
            .unwrap_or(0);

        match format {
            OutputFormat::Json => {
                println!("{{\"id\":\"{}\",\"file_count\":{}}}", id, files)
            }
            OutputFormat::Table => {
                println!("Imported snapshot {}", id);
                println!("  Files: {}", files);
                println!();
                println!("Apply it with: neonfs dr snapshot apply {}", id);
            }
        }

        Ok(())
    }

    fn apply(id: &str, format: OutputFormat) -> Result<()> {
        let id_term = Term::Binary(Binary {
            bytes: id.as_bytes().to_vec(),
        });

        let result = smol::block_on(async {
            let mut conn = DaemonConnection::connect().await?;
            conn.call(
                "Elixir.NeonFS.CLI.Handler",
                "handle_dr_snapshot_apply",
                vec![id_term],
            )
            .await
        })?;

        if let Some(err) = extract_error(&result) {
            return Err(err);
        }

        let restore = DrRestoreResult::from_term(unwrap_ok_tuple(result)?)?;

        match format {
            OutputFormat::Json => println!("{}", json::format(&restore)?),
            OutputFormat::Table => {
                println!("Applied snapshot {}", restore.id);
                println!("  Entries restored: {}", restore.total);
                println!("  Cluster generation: {}", restore.generation);
                for (keyspace, count) in &restore.restored {
                    println!("    {:<20} {}", keyspace, count);
                }
            }
        }

        Ok(())
    }

    fn create(label: Option<&str>, format: OutputFormat) -> Result<()> {
        let mut opts = vec![];

        if let Some(label) = label {
            opts.push((
                Term::Binary(Binary {
                    bytes: b"label".to_vec(),
                }),
                Term::Binary(Binary {
                    bytes: label.as_bytes().to_vec(),
                }),
            ));
        }

        let opts_term = Term::Map(Map {
            map: opts.into_iter().collect(),
        });

        let result = smol::block_on(async {
            let mut conn = DaemonConnection::connect().await?;
            conn.call(
                "Elixir.NeonFS.CLI.Handler",
                "handle_dr_snapshot_create",
                vec![opts_term],
            )
            .await
        })?;

        if let Some(err) = extract_error(&result) {
            return Err(err);
        }

        let entry = DrSnapshotEntry::from_term(unwrap_ok_tuple(result)?)?;

        match format {
            OutputFormat::Json => println!("{}", json::format(&entry)?),
            OutputFormat::Table => {
                println!("Created snapshot {}", entry.id);
                println!("  Path:          {}", entry.path);
                println!("  Created:       {}", entry.created_at);
                println!("  State version: {:?}", entry.state_version);
                println!("  Files:         {}", entry.file_count);
                println!("  Bytes:         {}", entry.total_bytes);
            }
        }

        Ok(())
    }

    fn list(format: OutputFormat) -> Result<()> {
        let result = smol::block_on(async {
            let mut conn = DaemonConnection::connect().await?;
            conn.call(
                "Elixir.NeonFS.CLI.Handler",
                "handle_dr_snapshot_list",
                vec![],
            )
            .await
        })?;

        if let Some(err) = extract_error(&result) {
            return Err(err);
        }

        let entry_terms = term_to_list(&unwrap_ok_tuple(result)?)?;
        let entries: Result<Vec<DrSnapshotEntry>> = entry_terms
            .into_iter()
            .map(DrSnapshotEntry::from_term)
            .collect();
        let entries = entries?;

        match format {
            OutputFormat::Json => println!("{}", json::format(&entries)?),
            OutputFormat::Table => {
                if entries.is_empty() {
                    println!("No DR snapshots found.");
                } else {
                    let mut tbl = table::Table::new(vec![
                        "ID".to_string(),
                        "CREATED".to_string(),
                        "STATE_VERSION".to_string(),
                        "FILES".to_string(),
                        "BYTES".to_string(),
                    ]);
                    for entry in &entries {
                        tbl.add_row(vec![
                            entry.id_short(),
                            entry.created_at.clone(),
                            entry
                                .state_version
                                .map(|v| v.to_string())
                                .unwrap_or_else(|| "-".to_string()),
                            entry.file_count.to_string(),
                            entry.total_bytes.to_string(),
                        ]);
                    }
                    print!("{}", tbl.render()?);
                }
            }
        }

        Ok(())
    }

    fn show(id: &str, format: OutputFormat) -> Result<()> {
        let id_term = Term::Binary(Binary {
            bytes: id.as_bytes().to_vec(),
        });

        let result = smol::block_on(async {
            let mut conn = DaemonConnection::connect().await?;
            conn.call(
                "Elixir.NeonFS.CLI.Handler",
                "handle_dr_snapshot_show",
                vec![id_term],
            )
            .await
        })?;

        if let Some(err) = extract_error(&result) {
            return Err(err);
        }

        let entry = DrSnapshotEntry::from_term(unwrap_ok_tuple(result)?)?;

        match format {
            OutputFormat::Json => println!("{}", json::format(&entry)?),
            OutputFormat::Table => {
                println!("Snapshot {}", entry.id);
                println!("  Path:          {}", entry.path);
                println!("  Created:       {}", entry.created_at);
                println!("  State version: {:?}", entry.state_version);
                println!("  Files:         {}", entry.file_count);
                println!("  Bytes:         {}", entry.total_bytes);
            }
        }

        Ok(())
    }
}

fn binary(s: &str) -> Term {
    Term::Binary(Binary {
        bytes: s.as_bytes().to_vec(),
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use clap::Parser;

    #[derive(Parser)]
    struct TestCli {
        #[command(subcommand)]
        command: DrCommand,
    }

    #[test]
    fn parses_snapshot_create_with_label() {
        let cli = TestCli::try_parse_from(["test", "snapshot", "create", "--label", "manual"]);
        assert!(cli.is_ok());
    }

    #[test]
    fn parses_snapshot_create_without_label() {
        let cli = TestCli::try_parse_from(["test", "snapshot", "create"]);
        assert!(cli.is_ok());
    }

    #[test]
    fn parses_snapshot_list() {
        let cli = TestCli::try_parse_from(["test", "snapshot", "list"]);
        assert!(cli.is_ok());
    }

    #[test]
    fn parses_snapshot_show() {
        let cli = TestCli::try_parse_from(["test", "snapshot", "show", "20260425T120000Z"]);
        assert!(cli.is_ok());
    }

    #[test]
    fn show_requires_id() {
        let cli = TestCli::try_parse_from(["test", "snapshot", "show"]);
        assert!(cli.is_err());
    }

    #[test]
    fn parses_snapshot_apply() {
        let cli = TestCli::try_parse_from(["test", "snapshot", "apply", "20260425T120000Z"]);
        assert!(cli.is_ok());
    }

    #[test]
    fn apply_requires_id() {
        let cli = TestCli::try_parse_from(["test", "snapshot", "apply"]);
        assert!(cli.is_err());
    }

    #[test]
    fn parses_snapshot_export() {
        let cli = TestCli::try_parse_from([
            "test",
            "snapshot",
            "export",
            "20260425T120000Z",
            "--out",
            "/tmp/dr",
        ]);
        assert!(cli.is_ok());
    }

    #[test]
    fn export_requires_out() {
        let cli = TestCli::try_parse_from(["test", "snapshot", "export", "20260425T120000Z"]);
        assert!(cli.is_err());
    }

    #[test]
    fn parses_snapshot_import() {
        let cli = TestCli::try_parse_from(["test", "snapshot", "import", "/tmp/dr"]);
        assert!(cli.is_ok());
    }

    #[test]
    fn parses_restore_with_source_only() {
        let cli = TestCli::try_parse_from(["test", "restore", "--source", "/tmp/dr"]);
        assert!(cli.is_ok());
    }

    #[test]
    fn parses_restore_with_catalogue_and_passphrase() {
        let cli = TestCli::try_parse_from([
            "test",
            "restore",
            "--source",
            "/tmp/dr",
            "--catalogue",
            "/tmp/dr/catalogue.json",
            "--passphrase",
            "hunter2",
        ]);
        assert!(cli.is_ok());
    }

    #[test]
    fn restore_requires_source() {
        let cli = TestCli::try_parse_from(["test", "restore"]);
        assert!(cli.is_err());
    }
}

File: neonfs_core/lib/neon_fs/cli/handler/dr.ex
defmodule NeonFS.CLI.Handler.DR do
  @moduledoc """
  CLI command handlers for disaster-recovery snapshots: create, list,
  and show the cluster-state DR snapshots in the `_system` volume's
  `/dr` directory (#324), plus the full-cluster `dr restore`
  orchestration (#1005).

  Extracted from `NeonFS.CLI.Handler` (#1203). `NeonFS.CLI.Handler`
  delegates its `handle_dr_snapshot_*` and `handle_dr_restore` RPC entry
  points here.
  """

  import NeonFS.CLI.Handler.Common

  alias NeonFS.Core.{AuditLog, Backup, DRSnapshot, VolumeRegistry}
  alias NeonFS.Error.NotFound

  @doc """
  Triggers an immediate DR snapshot (#324).
  """
  @spec handle_dr_snapshot_create(map()) :: {:ok, map()} | {:error, term()}
  def handle_dr_snapshot_create(_opts \\ %{}) do
    set_cli_metadata()

    with :ok <- require_cluster(),
         {:ok, snapshot} <- DRSnapshot.create([]) do
      AuditLog.log_event(
        event_type: :dr_snapshot_created,
        actor_uid: 0,
        resource: snapshot.path,
        details: %{
          state_version: snapshot.manifest.state_version,
          files: length(snapshot.manifest.files)
        }
      )

      {:ok, dr_snapshot_to_serialisable(snapshot)}
    else
      {:error, reason} -> {:error, wrap_error(reason)}
    end
  end

  @doc """
  Lists every DR snapshot, newest first (#324).
  """
  @spec handle_dr_snapshot_list() :: {:ok, [map()]} | {:error, term()}
  def handle_dr_snapshot_list do
    set_cli_metadata()

    with :ok <- require_cluster(),
         {:ok, snapshots} <- DRSnapshot.list() do
      {:ok, Enum.map(snapshots, &dr_snapshot_to_serialisable/1)}
    else
      {:error, reason} -> {:error, wrap_error(reason)}
    end
  end

  @doc """
  Fetches a single DR snapshot's manifest by id (#324).
  """
  @spec handle_dr_snapshot_show(String.t()) :: {:ok, map()} | {:error, term()}
  def handle_dr_snapshot_show(id) when is_binary(id) do
    set_cli_metadata()

    with :ok <- require_cluster(),
         {:ok, snapshot} <- DRSnapshot.get(id) do
      {:ok, dr_snapshot_to_serialisable(snapshot)}
    else
      {:error, :not_found} ->
        {:error, NotFound.exception(message: "DR snapshot '#{id}' not found")}

      {:error, reason} ->
        {:error, wrap_error(reason)}
    end
  end

  @doc """
  Apply a DR snapshot's cluster-wide metadata back into live Ra state
  (#1005) — the restore-primitive slice of full-cluster `dr restore`.
  """
  @spec handle_dr_snapshot_apply(String.t()) :: {:ok, map()} | {:error, term()}
  def handle_dr_snapshot_apply(id) when is_binary(id) do
    set_cli_metadata()

    with :ok <- require_cluster(),
         {:ok, %{restored: counts, generation: generation}} <- DRSnapshot.restore(id) do
      AuditLog.log_event(
        event_type: :dr_snapshot_applied,
        actor_uid: 0,
        resource: id,
        details: %{restored: counts, generation: generation}
      )

      {:ok,
       %{
         id: id,
         restored: counts,
         total: counts |> Map.values() |> Enum.sum(),
         generation: generation
       }}
    else
      {:error, :not_found} ->
        {:error, NotFound.exception(message: "DR snapshot '#{id}' not found")}

      {:error, reason} ->
        {:error, wrap_error(reason)}
    end
  end

  @doc """
  Export a DR snapshot off-cluster to `dest_dir` on the daemon's
  filesystem (#1367) so it survives a bare-metal disaster.
  """
  @spec handle_dr_snapshot_export(String.t(), String.t()) :: {:ok, map()} | {:error, term()}
  def handle_dr_snapshot_export(id, dest_dir) when is_binary(id) and is_binary(dest_dir) do
    set_cli_metadata()

    with :ok <- require_cluster(),
         {:ok, summary} <- DRSnapshot.export(id, dest_dir) do
      AuditLog.log_event(
        event_type: :dr_snapshot_exported,
        actor_uid: 0,
        resource: id,
        details: %{dest: summary.dest, file_count: summary.file_count}
      )

      {:ok, summary}
    else
      {:error, :not_found} ->
        {:error, NotFound.exception(message: "DR snapshot '#{id}' not found")}

      {:error, reason} ->
        {:error, wrap_error(reason)}
    end
  end

  @doc """
  Stage an exported DR snapshot from `source_dir` back into the
  `_system` volume (#1367) so `handle_dr_snapshot_apply/1` can consume
  it on a freshly-bootstrapped cluster.
  """
  @spec handle_dr_snapshot_import(String.t()) :: {:ok, map()} | {:error, term()}
  def handle_dr_snapshot_import(source_dir) when is_binary(source_dir) do
    set_cli_metadata()

    with :ok <- require_cluster(),
         {:ok, summary} <- DRSnapshot.import_snapshot(source_dir) do
      AuditLog.log_event(
        event_type: :dr_snapshot_imported,
        actor_uid: 0,
        resource: summary.id,
        details: %{source: source_dir, file_count: summary.file_count}
      )

      {:ok, summary}
    else
      {:error, reason} -> {:error, wrap_error(reason)}
    end
  end

  @doc """
  Full-cluster restore (#1005): stage the DR snapshot exported at
  `source_dir` back into the freshly-bootstrapped `_system` volume, apply
  its cluster-wide metadata (recreating every volume's shell and the
  cluster CA / credentials / keys), then restore each volume's content
  from its backup archive into that shell.

  This drives steps 2–3 of the bare-metal runbook; the operator does
  step 1 (`neonfs cluster init` on a fresh node) and step 4
  (`neonfs cluster join` for the remaining nodes) around it.

  `opts`:
    * `"catalogue"` — path to a JSON object `{"<volume>": "<archive>"}`
      pinning where each volume's backup archive lives. Relative archive
      paths resolve against `source_dir`; the catalogue is authoritative,
      so volumes it omits are left as empty shells. Without a catalogue,
      each volume's archive is expected at
      `<source_dir>/volumes/<volume>.backup`.
    * `"passphrase"` — passphrase for encrypted backup archives (#1004).

  Best-effort per volume: a volume whose archive is missing or fails to
  restore is reported in the result rather than aborting the whole
  restore, so one pass recovers everything recoverable.
  """
  @spec handle_dr_restore(String.t(), map()) :: {:ok, map()} | {:error, term()}
  def handle_dr_restore(source_dir, opts \\ %{}) when is_binary(source_dir) do
    set_cli_metadata()

    with :ok <- require_cluster(),
         {:ok, catalogue} <- load_catalogue(Map.get(opts, "catalogue"), source_dir),
         {:ok, %{id: snapshot_id}} <- DRSnapshot.import_snapshot(source_dir),
         {:ok, %{restored: counts, generation: generation}} <- DRSnapshot.restore(snapshot_id) do
      volume_results = restore_volumes(source_dir, catalogue, opts)

      AuditLog.log_event(
        event_type: :dr_restored,
        actor_uid: 0,
        resource: snapshot_id,
        details: %{generation: generation, volumes: length(volume_results)}
      )

      {:ok,
       %{
         snapshot_id: snapshot_id,
         generation: generation,
         restored: counts,
         volumes: volume_results,
         volumes_restored: Enum.count(volume_results, &(&1.status == "restored")),
         volumes_failed: Enum.count(volume_results, &(&1.status == "failed"))
       }}
    else
      {:error, reason} -> {:error, wrap_error(reason)}
    end
  end

  # Private

  defp restore_volumes(source_dir, catalogue, opts) do
    restore_opts = backup_restore_opts(opts)

    VolumeRegistry.list()
    |> Enum.map(fn volume ->
      case archive_for(volume.name, catalogue, source_dir) do
        {:ok, archive} -> restore_one_volume(volume.name, archive, restore_opts)
        :none -> %{name: volume.name, status: "skipped", reason: "no backup archive"}
      end
    end)
  end

  defp restore_one_volume(name, archive, restore_opts) do
    case Backup.restore(archive, name, [{:into_existing, true} | restore_opts]) do
      {:ok, summary} ->
        %{
          name: name,
          status: "restored",
          archive: archive,
          files: summary.file_count,
          bytes: summary.byte_count
        }

      {:error, reason} ->
        %{name: name, status: "failed", archive: archive, reason: inspect(reason)}
    end
  end

  defp archive_for(name, catalogue, _source_dir) when is_map(catalogue) do
    case Map.fetch(catalogue, name) do
      {:ok, archive} -> {:ok, archive}
      :error -> :none
    end
  end

  defp archive_for(name, nil, source_dir) do
    archive = Path.join([source_dir, "volumes", "#{name}.backup"])
    if File.exists?(archive), do: {:ok, archive}, else: :none
  end

  defp load_catalogue(nil, _source_dir), do: {:ok, nil}

  defp load_catalogue(path, source_dir) when is_binary(path) do
    with {:ok, raw} <- File.read(path),
         {:ok, map} when is_map(map) <- Jason.decode(raw) do
      {:ok, Map.new(map, fn {name, archive} -> {name, resolve_archive(archive, source_dir)} end)}
    else
      {:ok, _non_object} ->
        {:error, {:invalid_catalogue, "catalogue must be a JSON object of volume => archive"}}

      {:error, %Jason.DecodeError{}} ->
        {:error, {:invalid_catalogue, "catalogue is not valid JSON"}}

      {:error, reason} ->
        {:error, reason}
    end
  end

  defp resolve_archive(archive, source_dir) do
    if Path.type(archive) == :absolute, do: archive, else: Path.join(source_dir, archive)
  end

  defp backup_restore_opts(opts) do
    case Map.get(opts, "passphrase") do
      passphrase when is_binary(passphrase) and passphrase != "" -> [passphrase: passphrase]
      _ -> []
    end
  end

  defp dr_snapshot_to_serialisable(%{id: id, path: path, manifest: manifest}) do
    %{
      id: id,
      path: path,
      version: manifest.version,
      created_at: manifest.created_at,
      state_version: manifest.state_version,
      file_count: length(manifest.files || []),
      total_bytes: total_manifest_bytes(manifest.files || []),
      files: Enum.map(manifest.files || [], &dr_snapshot_file_to_serialisable/1)
    }
  end

  defp dr_snapshot_to_serialisable(%{path: path, manifest: manifest}) do
    # `DRSnapshot.create/1` returns `{:ok, %{path, manifest}}` without an
    # id key; derive the id from the directory name so the CLI surface
    # is uniform across create / list / show.
    id = Path.basename(path)
    dr_snapshot_to_serialisable(%{id: id, path: path, manifest: manifest})
  end

  defp dr_snapshot_file_to_serialisable(file) do
    %{
      path: Map.get(file, :path),
      bytes: Map.get(file, :bytes),
      sha256: Map.get(file, :sha256),
      kind: file |> Map.get(:kind) |> to_string()
    }
  end

  defp total_manifest_bytes(files) do
    Enum.reduce(files, 0, fn f, acc -> acc + (Map.get(f, :bytes) || 0) end)
  end
end
